{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"source":"<a href=\"https://www.kaggle.com/code/mohdmuttalib/novozymes-enzyme-stability-prediction?scriptVersionId=145004971\" target=\"_blank\"><img align=\"left\" alt=\"Kaggle\" title=\"Open in Kaggle\" src=\"https://kaggle.com/static/images/open-in-kaggle.svg\"></a>","metadata":{},"cell_type":"markdown"},{"cell_type":"markdown","source":"## Contact Information\n\n- **Name** - Mohd Muttalib\n- **Phone** - +91-8445818187\n- **Email** - muttalib1326@gmail.com\n- **Portfolio** -[link](https://www.kaggle.com/mohdmuttalib)","metadata":{}},{"cell_type":"markdown","source":"<center><img src=\"https://storage.googleapis.com/kaggle-competitions/kaggle/37190/logos/header.png?t=2022-08-30-15-34-26\" width=1000></center>\n\n\n\n<div class=\"alert alert-block alert-info\">\n<center><h1>Novozymes Enzyme Stability Prediction</h1></center>\n<center><h3>Help identify the thermostable mutations in enzymes</h3></center>\n</div>\n<div class=\"alert alert-block alert-success\">\n<center><h1>EDA + XGBoost</h1></center>\n</div>\n\n\n### <h1>Table of contents</h1><a class='anchor' id='top'></a>\n- [Import libraries](#import)\n- [Utils](#utils)\n- [Config](#config)\n- [Load data](#load)\n- [Train profiling](#train_profiling)\n- [Test profiling](#test_profiling)\n- [What is PDB?](#pdb)\n- [Label Encode](#labelencode)\n- [XGBoost Regressor](#xgboost)\n- [Evaluate: Spearman's rank correlation coefficient](#evaluate)\n- [Predict](#predict)\n- [Submission](#submission)\n\n\n<div class=\"alert alert-block alert-warning\">  \n<b>Note:</b> This is a work in progress ⚠️\n</div>\n\n<div class=\"alert alert-block alert-info\">\n<center><h1>What is the goal of this competition?</h1></center>\n</div>\n\n<div class=\"alert alert-block alert-info\">\nThe goal of this competition is to predict the thermal stability of enzyme variants. The experimentally measured thermal stability (melting temperature) data includes natural sequences, as well as engineered sequences with single or multiple mutations upon the natural sequences.\n\nA point mutation is a genetic mutation where a single nucleotide base is changed, inserted or deleted from a DNA or RNA sequence of an organism's genome. Point mutations have a variety of effects on the downstream protein product—consequences that are moderately predictable based upon the specifics of the mutation.\n\nThese consequences can range from no effect (e.g. synonymous mutations) to deleterious effects (e.g. frameshift mutations), with regard to protein production, composition, and function.\n</div>\n\n\n\n\n<div class=\"alert alert-block alert-success\">\n<center><h1>What is an enzyme?</h1></center>\n</div>\n<div class=\"alert alert-block alert-success\">\nEnzymes are proteins that act as biological catalysts by accelerating chemical reactions. An enzyme is therefore a molecule that increases the speed of biochemical reaction in an organism. The molecules upon which enzymes may act are called substrates, and the enzyme converts the substrates into different molecules known as products.\n</div>\n<center><img src=\"https://www.googleapis.com/download/storage/v1/b/kaggle-forum-message-attachments/o/inbox%2F3197853%2F011e8e6d88265aaf8d13c7b9cae4b2b8%2Fenzyme_substrate.jpeg?generation=1663958942493732&alt=media\" width=800></center>\n<div class=\"alert alert-block alert-success\">\n<center><h1>What is a protein?</h1></center>\n</div>\n<div class=\"alert alert-block alert-success\">\nProteins are large biomolecules and macromolecules that comprise one or more long chains of amino acid residues. Proteins are also known as polypeptides.\n\nPeptides are short chains of amino acids linked by peptide bonds. Chains of fewer than twenty amino acids are called oligopeptides, and include dipeptides, tripeptides, and tetrapeptides. So, a polypeptide is a longer, continuous, unbranched peptide chain.\n</div>\n<center><img src=\"https://www.googleapis.com/download/storage/v1/b/kaggle-forum-message-attachments/o/inbox%2F3197853%2F649adf375f36aaf3cad5064cf8d543eb%2Fprotein_structure.jpeg?generation=1663958961249770&alt=media\" width=600></center>\n<div class=\"alert alert-block alert-success\">\n<center><h1>What is an amino acid?</h1></center>\n</div>\n<div class=\"alert alert-block alert-success\">\nAmino acids are organic compounds that contain both amino and carboxylic acid functional groups. Although hundreds of amino acids exist in nature, by far the most prevalent are the alpha-amino acids, which comprise proteins.\n</div>\n<center><img src=\"https://www.googleapis.com/download/storage/v1/b/kaggle-forum-message-attachments/o/inbox%2F3197853%2Ff6f63060a29aa51e3675e36e27c51b9d%2Faminoacids.png?generation=1663958987152864&alt=media\" width=600></center>\n\n","metadata":{}},{"cell_type":"code","source":"!pip install biopandas","metadata":{"_kg_hide-output":true,"scrolled":true,"_kg_hide-input":true,"execution":{"iopub.status.busy":"2023-10-02T17:58:21.903996Z","iopub.execute_input":"2023-10-02T17:58:21.904527Z","iopub.status.idle":"2023-10-02T17:58:31.689765Z","shell.execute_reply.started":"2023-10-02T17:58:21.904488Z","shell.execute_reply":"2023-10-02T17:58:31.688461Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Import libraries <a class='anchor' id='import'></a> [↑](#top)","metadata":{}},{"cell_type":"code","source":"import matplotlib.pyplot as plt\nimport numpy as np\nimport os\nimport pandas as pd\nimport pandas_profiling\nimport time\nimport torch\nimport torch.nn as nn","metadata":{"execution":{"iopub.status.busy":"2023-10-02T17:58:31.693243Z","iopub.execute_input":"2023-10-02T17:58:31.693818Z","iopub.status.idle":"2023-10-02T17:58:31.701179Z","shell.execute_reply.started":"2023-10-02T17:58:31.693762Z","shell.execute_reply":"2023-10-02T17:58:31.700053Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Utils <a class='anchor' id='utils'></a>[↑](#top)","metadata":{}},{"cell_type":"code","source":"def sep():\n    print(\"-\"*100)","metadata":{"execution":{"iopub.status.busy":"2023-10-02T17:58:31.702748Z","iopub.execute_input":"2023-10-02T17:58:31.703098Z","iopub.status.idle":"2023-10-02T17:58:31.723511Z","shell.execute_reply.started":"2023-10-02T17:58:31.703065Z","shell.execute_reply":"2023-10-02T17:58:31.722363Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Config <a class='anchor' id='config'></a>[↑](#top)","metadata":{}},{"cell_type":"code","source":"class paths:\n    TRAIN = \"/kaggle/input/novozymes-enzyme-stability-prediction/train.csv\"\n    TEST = \"/kaggle/input/novozymes-enzyme-stability-prediction/test.csv\"\n    SUBMISSION = \"/kaggle/input/novozymes-enzyme-stability-prediction/sample_submission.csv\"\n    PDB_FILE = \"/kaggle/input/novozymes-enzyme-stability-prediction/wildtype_structure_prediction_af2.pdb\"","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","execution":{"iopub.status.busy":"2023-10-02T17:58:31.72635Z","iopub.execute_input":"2023-10-02T17:58:31.72675Z","iopub.status.idle":"2023-10-02T17:58:31.738626Z","shell.execute_reply.started":"2023-10-02T17:58:31.72671Z","shell.execute_reply":"2023-10-02T17:58:31.737031Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Load data <a class='anchor' id='load'></a>[↑](#top)\n\nLet's load the competition's data.","metadata":{}},{"cell_type":"code","source":"train_df = pd.read_csv(paths.TRAIN)\ntest_df = pd.read_csv(paths.TEST)\nprint(f\"Train dataframe has shape: {train_df.shape}\")\nprint(f\"Test dataframe has shape: {test_df.shape}\")\ndisplay(train_df.head())\ndisplay(test_df.head())","metadata":{"execution":{"iopub.status.busy":"2023-10-02T17:58:31.740563Z","iopub.execute_input":"2023-10-02T17:58:31.741038Z","iopub.status.idle":"2023-10-02T17:58:32.017211Z","shell.execute_reply.started":"2023-10-02T17:58:31.741003Z","shell.execute_reply":"2023-10-02T17:58:32.015782Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Train profiling <a class='anchor' id='train_profiling'></a>[↑](#top)\n\n- `protein_sequence`: has a high cardinality: 28981 distinct values ---> High cardinality\n- `data_source`: has a high cardinality: 324 distinct values ---> High cardinality\n- `data_source` has 3347 (10.7%) missing values ---> Missing\n- `protein_sequence_len` is highly skewed (γ1 = 29.89928625) ---> Skewed\n- `seq_id` is uniformly distributed ---> Uniform\n- `protein_sequence` is uniformly distributed ---> Uniform\n- `seq_id` has unique values","metadata":{}},{"cell_type":"code","source":"train_df.profile_report()","metadata":{"execution":{"iopub.status.busy":"2023-10-02T17:58:32.0188Z","iopub.execute_input":"2023-10-02T17:58:32.019141Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Test profiling <a class='anchor' id='test_profiling'></a>[↑](#top)\n\n- `pH` has constant value \"8\" ---> Constant\n- `data_source` has constant value \"Novozymes\" ---> Constant\n- `protein_sequence` has a high cardinality: 2413 distinct values ---> High cardinality\n- `pH` is highly correlated with data_source and 1 other fields ---> High correlation\n- `data_source` is highly correlated with pH and 1 other fields ---> High correlation\n- `protein_sequence_len` is highly correlated with pH and 1 other fields ---> High correlation\n- `seq_id` is uniformly distributed ---> Uniform\n- `protein_sequence` is uniformly distributed ---> Uniform\n- `seq_id` has unique values ---> Unique\n- `protein_sequence` has unique values","metadata":{}},{"cell_type":"code","source":"test_df.profile_report()","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Plot target [↑](#top)\n\nLet's plot our target so we can get a sense of its distribution. We can see that it's right skewed with a large value around 25.","metadata":{}},{"cell_type":"code","source":"from plotly.subplots import make_subplots\nimport plotly.graph_objects as go\n\n\nfig = go.Figure()\ncolors = [\"#00FFFB\", \"#002EFF\"]\n\ni = 0\n# Add one subplot\nfig.add_trace(\n    go.Histogram(x=train_df[\"tm\"],\n               name=\"tm\",\n               hovertemplate = \"target column\" + ' %{y:.2f}',\n               marker=dict(color=colors[i])\n              )\n)\nfig.update_xaxes(\n    title_text=\"Target column\",\n    title_font_color=colors[i],\n    tickfont_color=colors[i]) \nfig.update_yaxes(\n    title_text = \"Frequency\", \n    title_font_color=colors[i], \n    tickfont_color=colors[i]) \n    \nfig.update_layout(height=800,\n                  width=1000,\n                  title_text=\"Target column: Higher tm means the protein variant is more stable.\",\n                  template=\"plotly_dark\",\n                  xaxis=dict(color=\"#FF9300\"),\n                  yaxis=dict(color=\"#FF9300\"))\n\nfig.update_traces(marker_line_width=0.1,marker_line_color=colors[1])\nfig.show()","metadata":{"_kg_hide-input":true,"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Sequence length [↑](#top)\n\nSequence length seems to vary for the `train.csv` while it is constant for `test.csv`\n\nWe can see there is one heavy outlier of length `32767`. The distribution is highly right skewed.","metadata":{}},{"cell_type":"code","source":"train_df[\"protein_sequence_len\"] = train_df[\"protein_sequence\"].apply(lambda x: len(x))\ntest_df[\"protein_sequence_len\"] = test_df[\"protein_sequence\"].apply(lambda x: len(x))\nprint(\"----------- TRAIN -----------\")\ndisplay(train_df[[\"protein_sequence_len\"]].describe())\nprint(\"----------- TEST -----------\")\ndisplay(test_df[[\"protein_sequence_len\"]].describe())","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"from plotly.subplots import make_subplots\nimport plotly.graph_objects as go\n\n\nfig = go.Figure()\ncolors = [\"#FF00F0\", \"#BD00FF\"]\n\ni = 0\n# Add one subplot\nfig.add_trace(\n    go.Histogram(x=train_df[\"protein_sequence_len\"],\n               name=\"protein_sequence_len\",\n               hovertemplate = 'Protein Sequence Length:  %{x:.2f}' + '<br>Frequency: %{y:.2f}</br>',\n               marker=dict(color=colors[i])\n              )\n)\nfig.update_xaxes(\n    title_text=\"Protein Sequence Length\",\n    title_font_color=colors[i],\n    tickfont_color=colors[i],\n    range=[0, 1500]\n) \nfig.update_yaxes(\n    title_text = \"Frequency\", \n    title_font_color=colors[i], \n    tickfont_color=colors[i]\n) \n\nfig.update_layout(height=800,\n                  width=1000,\n                  title_text=\"Potein Sequence Length Histogram.\",\n                  template=\"plotly_dark\",\n                  xaxis=dict(color=\"#FF9300\"),\n                  yaxis=dict(color=\"#FF9300\"))\nfig.update_traces(marker_line_width=2,marker_line_color=colors[1])\n\nfig.show()","metadata":{"_kg_hide-input":true,"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# What is PDB? <a class='anchor' id='pdb'></a>[↑](#top)\n\nPDB file format\n\nIn the PDB data file format for macromolecular models, each atom is designated either ATOM or HETATM (which stands for hetero atom).\n\nATOM is reserved for atoms in standard residues of protein, DNA or RNA.\n\nHETATM is applied to non-standard residues of protein, DNA or RNA, as well as atoms in other kinds of groups, such as carbohydrates, substrates, ligands, solvent, and metal ions.","metadata":{}},{"cell_type":"code","source":"from biopandas.pdb import PandasPdb\n\npdb_df =  PandasPdb().read_pdb(paths.PDB_FILE)\npdb_df.df.keys()","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Get dataframes\n\nThe `pdb` file contains data which we can convert into new dataframes. Let's check it out.","metadata":{}},{"cell_type":"code","source":"atom_df = pdb_df.df['ATOM']\nhetatm_df = pdb_df.df['HETATM']\nanisou_df = pdb_df.df['ANISOU']\nothers_df = pdb_df.df['OTHERS']\nprint(f\"ATOM dataset is of shape: {atom_df.shape}\"), sep()\nprint(f\"HETATM dataset is of shape: {hetatm_df.shape}\"), sep()\nprint(f\"ANISOU dataset is of shape: {anisou_df.shape}\"), sep()\nprint(f\"OTHERS dataset is of shape: {others_df.shape}\"), sep()\ndisplay(atom_df.head())\ndisplay(hetatm_df.head())\ndisplay(anisou_df.head())\ndisplay(others_df.head())","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Plot sequence","metadata":{}},{"cell_type":"code","source":"import plotly.express as px\n\nfig = px.scatter_3d(atom_df, x = \"x_coord\",\n                    y = \"y_coord\",\n                    z = \"z_coord\",\n                    color = \"element_symbol\",\n                    color_discrete_sequence = [\"#84FFA9\", \"#00FFF7\", \"#003AFF\", \"#F000FF\", \"#FBFF00\"])\nfig.update_traces(marker = dict(size = 3))\nfig.update_coloraxes(showscale = False)\nfig.update_layout(template = \"plotly_dark\")\nfig.show()","metadata":{"_kg_hide-input":true,"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Get sequences [↑](#top)\n\nWe will only use sequences lower than 221.","metadata":{}},{"cell_type":"code","source":"from scipy.sparse import csr_matrix\n\ntrain_df = train_df[train_df[\"protein_sequence_len\"]<=221]\ntrain_df.reset_index(inplace=True)\nsequences = [list(string) for string in train_df[\"protein_sequence\"].values.tolist()]\nsequences_train = pd.DataFrame(sequences)\nsequences_train.head()","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Label encode <a class='anchor' id='labelencode'></a>[↑](#top)\n\nWe need to label encode data so we can use XGBoost Regressor","metadata":{}},{"cell_type":"code","source":"from sklearn.preprocessing import LabelEncoder\n\nsequences_train = sequences_train.apply(LabelEncoder().fit_transform)\nsequences_train[\"tm\"] = train_df[\"tm\"]\nsequences_train.head()","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# XGBoost Regressor <a class='anchor' id='xgboost'></a>[↑](#top)\n\nLet's make a 80/20 split and train a XGBoost regressor.","metadata":{}},{"cell_type":"code","source":"from sklearn.model_selection import train_test_split\nimport xgboost\n\nX = sequences_train.loc[:, sequences_train.columns != \"tm\"]\ny = sequences_train.loc[:, sequences_train.columns == \"tm\"]\n\nX_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n# create an xgboost regression model\nmodel = xgboost.XGBRegressor(n_estimators=500, max_depth=15)\nmodel.fit(X_train, y_train)\ny_pred = model.predict(X_test)","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Evaluate: Spearman's rank correlation coefficient <a class='anchor' id='evaluate'></a>[↑](#top)\n\nWe need to evaluate with [Spearman's rank correlation coefficient](https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient)","metadata":{}},{"cell_type":"code","source":"from scipy import stats\n\nstats.spearmanr(y_test, y_pred)","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Predict <a class='anchor' id='predict'></a>[↑](#top)\n\nNow that the model is trained let's convert the `test` dataframe into the required format and get predictions.","metadata":{}},{"cell_type":"code","source":"from scipy.sparse import csr_matrix\n\ntest_df = test_df[test_df[\"protein_sequence_len\"]<=221]\nsequences = [list(string) for string in test_df[\"protein_sequence\"].values.tolist()]\nsequences_test = pd.DataFrame(sequences)\nsequences_test = sequences_test.apply(LabelEncoder().fit_transform)\nsequences_test.head()","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Submission <a class='anchor' id='submission'></a>[↑](#top)\n\nFinally, let's submit our predictions.","metadata":{}},{"cell_type":"code","source":"submission = pd.DataFrame()\nsubmission[\"tm\"] = model.predict(sequences_test)\nsubmission[\"seq_id\"] = test_df[\"seq_id\"]\nsubmission.to_csv(\"submission.csv\", index=False)","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}]}